### ISSUE TITLE
false positive edge-case of Missing top-level documentation comment on class split among several files

### ISSUE THREAD

**Issue by noraj** [rubocop/rubocop#13014]
I checked most issues for this cop `Style/Documentation: Missing top-level documentation comment` : https://github.com/rubocop/rubocop/issues?q=is%3Aissue+Style%2FDocumentation%3A+Missing+top-level+documentation+comment

## Context

Let's say I have a file `cli.rb`

```ruby
require 'cli/dns'

# top-level class comment
class CLI
  # some global stuff here
end
```

and a file `cli/dns.rb`

```ruby
class CLI
  # some stuff to DNS here
end
```

So I have the same class split among several files for better management over large code base.

## Expected behavior

I expect to have to comment the same class only once.

## Actual behavior

Rubocop detects that the class is commented on one file and so doesn't complain about it but complains that the same class is undocumented on another file. But since it's the same class I don't want to have to copy/paste the same documentation over several files, which is unpractical and prone to errors for maintenance.

## Steps to reproduce the problem

The bug doesn't appear in the minimal example I gave in the context section. So I don't know which edge case make this false positive appears in my real code base which seems extremly similar.

![image](https://github.com/rubocop/rubocop/assets/16578570/26fd306d-76c9-483f-a932-1a68feeed7fd)

![image](https://github.com/rubocop/rubocop/assets/16578570/401917bf-fdea-4517-ae8a-459233202c95)

It doesn't seem related to fact the class is in a module.

## RuboCop version

```
➜ bundle exec rubocop -V
1.64.1 (using Parser 3.3.2.0, rubocop-ast 1.31.3, running on ruby 3.3.0) [x86_64-linux]
```

**Comment by noraj** [comment]
Which is even stranger is that if I add `# rubocop:disable Style/Documentation` to suppress this cop on `cli/dns.rb` to avoid the false positive, it triggers `Lint/RedundantCopDisableDirective` like if there was no issue in the first place.

Without `# rubocop:disable Style/Documentation`

![image](https://github.com/rubocop/rubocop/assets/16578570/b38b663b-3994-4c33-ba07-adc86077c41d)

With `# rubocop:disable Style/Documentation`

![image](https://github.com/rubocop/rubocop/assets/16578570/5c31f43a-073f-4a3d-8d5d-0823e416aed4)

**Comment by Earlopain** [comment]
RuboCop wants you to reenable the cop you disabled somewhere, that is the second message you are getting. I'm not seeing something about redundant disable messages in your given example.

As for your actual issue, not much that rubocop can do. It has no concept of the bigger picture. I'd just put this in a module to include in the main class, seems neater anyways.

**Comment by noraj** [comment]
> RuboCop wants you to reenable the cop you disabled somewhere, that is the second message you are getting. I'm not seeing something about redundant disable messages in your given example.

My bad, wrong image, I can't reproduce the `Lint/RedundantCopDisableDirective` now.

> As for your actual issue, not much that rubocop can do. It has no concept of the bigger picture.

So Rubocop's analyze code only on a per-file basis, right?

> I'd just put this in a module to include in the main class, seems neater anyways.

So you mean, instead of splitting a class among several files, puts the extra content in a module, and include that module in the class, so there is only one file with a class declaration?

But actually it's code for [gli](https://github.com/davetron5000/gli) option parsing (commands, sub-commands, global option, local options, etc.), and the way gli is, everything is on the same level / scope, declared sequentially in blocks, there are no methods or classes to split that properly (eg like in [dry-cli](https://dry-rb.org/gems/dry-cli/1.0/)). But afaik, includes or extends require there methods to be imported as instance or class methods in the target class.

**Comment by Earlopain** [comment]
Yes, RuboCop in essense looks at each file in isolation. If you can't make separation like that work with that lib then I would just exclude your folder with all the commands from `Style/Documentation`.

**Comment by noraj** [comment]
> Yes, RuboCop in essense looks at each file in isolation. If you can't make separation like that work with that lib then I would just exclude your folder with all the commands from `Style/Documentation`.

Yeah that's what I did. So I guess there is no point of keeping that issue open then?
